Macos support.

covers1624 opened the pull request:
Implements cross-platform and macos patching.

macOS requires [Quill](https://github.com/anchore/quill) or Xcode cli tools to be available in order to ad-hoc sign the patched binary.

covers1624 committed:
Macos patching support.

covers1624 committed:
Unify all patching paths.

Chicken-Bones commented on Program.cs:
```diff
@@ -42,39 +29,14 @@
 	Console.WriteLine("Done");
 }
 catch (ArgumentException ex) {
-	Console.Error.WriteLine(ex.Message);
```
This was intentional. Otherwise there's no point separating out `ArgumentException`. I used it to provide user facing error messages

covers1624 replied:
Don't think I intended to commit this. Reverted.

Chicken-Bones commented on Patches.cs:
```diff
@@ -88,5 +88,96 @@ public static class Patches
+			[Architecture.Arm64] = Mac_aarch64
+		};
+
+		public static Dictionary<OSPlatform, Dictionary<Architecture, Patch[]>> PlatformPatchSets = new() {
```
Do you need to expose each of the `<Platform>PatchSets` dicts above? I think it looks better to create them inline in this nested dictionary here.

covers1624 replied:
Fixed, Merged all the dicts together.

covers1624 committed:
Rename Patcher to AssemblyProvider.

covers1624 committed:
Switch to MachoSharp for macOS.

covers1624 committed:
Add an optional output path argument.

Very useful for testing and not continually needing to replace my binary with a backup.

covers1624 committed:
Add ad-hoc signing for macos, using codesign or quill.

Implements ad-hoc signing (non-cryptographic) of the macos executable after patching. To implement this we either require `codesign` from the Xcode cli tools, or Quill (available on all platforms).

covers1624 committed:
Paths may have spaces.

covers1624 committed:
Fix codesign detection.

Codesign doesn't have a --help command or something we can easily test for exit code 0, so we exec and check for the start of the help message.

covers1624 committed:
Some minor cleanup.

Chicken-Bones commented on Program.cs:
```diff
@@ -1,38 +1,33 @@
 try {
 	if (args.Length <= 0)
-		throw new ArgumentException($"Usage: <path to factorio.exe or map.zip");
+		throw new ArgumentException($"Usage: <path to factorio.exe or map.zip> [<path to output>]");
```
`map.zip` approach got abandoned

covers1624 replied:
Nuked.

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+    public bool PreCheck() {
+        if (_signer == null) {
+            Console.WriteLine("[Error]: Unable to ad-hoc sign macos executable on this system.");
+            Console.WriteLine("         \"codesign\" form the Xcode command line tools, or Quill(https://github.com/anchore/quill)");
```
`form` -> `from`

covers1624 replied:
Fixed.

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+        if (_signer == null) {
+            Console.WriteLine("[Error]: Unable to ad-hoc sign macos executable on this system.");
+            Console.WriteLine("         \"codesign\" form the Xcode command line tools, or Quill(https://github.com/anchore/quill)");
+            Console.WriteLine("         is required to be available on PATH to patch macOS binaries.");
```
remove `to be available`

covers1624 replied:
Fixed.

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+            ProcessStartInfo start = new ProcessStartInfo {
+                FileName = name,
+                Arguments = args,
+                UseShellExecute = false,
```
this is the default value

covers1624 replied:
I blame tML for this, as I used it as reference for how to exec stuff :)

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+            FileName = "quill",
+            Arguments = "sign \"" + Path.GetFullPath(filePath) + "\"",
+            UseShellExecute = false,
+            RedirectStandardError = false,
+            RedirectStandardOutput = false,
```
These are the default values

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+    }
+    
+    
+    public bool PreCheck() {
```
Can we just throw an exception from `FindSignerProgram` instead of having a separate `PreCheck`? It can be of another type which only shows message and not stack trace in the top level catch. Also means `_signer` is no longer nullable

covers1624 replied:
Done.

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+public sealed class MacosAssemblyProvider : AssemblyProvider {
+
+    private Dictionary<Architecture, MachO64> binaries;
+    private Signer?                           _signer;
```
haha, I won't make you change it, but this spacing is very non-standard

covers1624 replied:
Rider default formatting rules aligns things. Have disabled it and formatted the file.

Chicken-Bones commented on AssemblyProvider.cs:
```diff
@@ -0,0 +1,274 @@
+            return new LinuxAssemblyProvider(elf);
+        }
+
+        if (MachOReader.TryLoadFat(new MemoryStream(moduleBytes), out var fat, false)) {
```
`fat` is not disposable here?

covers1624 replied:
Nope, MachOFat doesn't keep the stream open, or use any lazy loading. It pulls the model for the macho and maintains offsets to where it got it from in the underlying stream.

Chicken-Bones reviewed (changes requested):
Looks great, just code style change requests

covers1624 committed:
Address review comments.

Chicken-Bones commented on Program.cs:
```diff
@@ -48,6 +45,12 @@
 	Console.Error.WriteLine(ex.Message);
 }
+catch (SignerNotFoundException) {
+	Console.WriteLine("[Error]: Unable to ad-hoc sign macos executable on this system.");
+	Console.WriteLine("         \"codesign\" from the Xcode command line tools, or Quill(https://github.com/anchore/quill)");
+	Console.WriteLine("         is required on PATH to patch macOS binaries.");
```
Could reduce coupling if this was the exception message, and we just print `ex.Message`, but doesn't really matter

covers1624 committed:
Move signing error to exception message.
